TypeCobol.Analysis/Graph/CfgAbstractInterpretation.Environment.cs

```diff
@@ -0,0 +1,368 @@
+                    _cyclicThreshold = new Dictionary<int, int>();
+                else
+                    _cyclicThreshold.Clear();
+                if (_cyclicTransition == null)
+                    _cyclicTransition = new Dictionary<int, HashSet<int>>();
+                else
+                    _cyclicTransition.Clear();
+            }
+
+            /// <summary>
+            /// Run un abstract interpretation on a graph.
+            /// </summary>
+            /// <param name="cfg">The graph to be interpreted.</param>
+            /// <param name="cyclicExecutionThreshold">The maximal count of cyclic execution for a cyclic block.</param>
+            /// <returns>Execution metrics.</returns>
+            public virtual Metrics Run(ControlFlowGraph<N, D> cfg, int cyclicExecutionThreshold = 0)
+            {
+                this.Cfg = cfg;
+                this._cyclicExecutionThreshold = cyclicExecutionThreshold;
+                Reset();
+                SetCyclicityThreshold();
+                this._metrics = new Metrics();                
+                Run(cfg.RootBlock, null);
+                System.Diagnostics.Debug.Assert(_interpretationStack.Count == 0);
+                System.Diagnostics.Debug.Assert(_dfsStack.Count == 0);
+                return this._metrics;
+            }
+
+            /// <summary>
+            /// Set the Cyclicity Threshold for cyclic blocks of the graph.
+            /// </summary>
+            public void SetCyclicityThreshold()
+            {
+                BitSet _domain = new BitSet();
+                Dictionary<int, int> _N = new Dictionary<int, int>();
+                Stack<int> _stack = new Stack<int>();
+
+                //Compute the domain
+                Cfg.DFS((block, incomingEdge, predecessorBlock, graph) => { _domain.Set(block.Index); return true; });
+                for (int i = _domain.NextSetBit(0); i >= 0; i = _domain.NextSetBit(i + 1))
+                {
+                    if (Visited(i) == 0)
+                        dfs(i);
+                }
+
+                // Visit mark of a block 'a'
+                // 0 : Block not visited
+                // > 0 : Block in active consideration
+                // Infinity : the cycle to which belongs the block has been considered.
+                int Visited(int a)
+                {
+                    if (_N.TryGetValue(a, out var na))
+                        return na;
+                    return 0;
+                }
+
+                // Travserse of all transition from block 'a'
+                void dfs(int a)
+                {
+                    _stack.Push(a);
+                    int d = _stack.Count;
+                    _N[a] = d;
+                    foreach (var s in Cfg.AllBlocks[a].SuccessorEdges)
+                    {   // For each transition from block 'a' to block 'b'
+                        var sb = Cfg.SuccessorEdges[s];
+                        int b = sb.Index;
```

Comment:
Can you use more meaningful variable name.
Like:
```suggestion
                        int blockBIndex = Cfg.SuccessorEdges[s].Index;
```

Reply:
Changed in 98d7aaa